rfcs/2025-shadow-dom-support.md

```diff
@@ -0,0 +1,128 @@
+
+```jsx
+document.activeElement
+<->
+getActiveElement();
+
+e.target
+<->
+getTargetElement(e);
+
+let walker = document.createTreeWalker(...)
+<->
+let walker = createShadowTreeWalker(...)
+
+e.currentTarget.contains(e.target)
+<->
+nodeContains(e.currentTarget, getEventTarget(e.nativeEvent))
+```
+
+4. Communication
+
+We have our current Shadow DOM support gated behind a feature flag to both signal that it is experimental as well as to protect users who are not using any Shadow DOM from possible performance hits.
+
+We will want to expand this to also shield those users from code size increases. This could be accomplished with a global registry for the utility functions we provide that would replace the feature flag system we have for it right now.
+
+We also need to communicate, through documentation, what the specifics are of our Shadow DOM support since we likely will not be able to support everything, and certainly not much for a while.
+
+For example, some of our support may require open root Shadow DOMs. Or someone using our FocusScope's focus manager may need to know about different traversal methods to avoid the same issues we are trying to avoid.
+
+## Drawbacks
+
+One concern is on-going support. Our current use cases do not call for Shadow DOM support, so knowledge on the team is currently thin. It will take time for people to ramp up on skills. Not only that, but it isn't part of our weekly testing and we have no plans for it to be at this time, which means we must rely on unit tests as much as possible.
+
+In addition, contributions may occur for a little while until those teams have their needs met. However, it should be assumed that there will be further work to complete the goals as outlined here.
+
+Another concern is that the current approach is, for lack of a better word, hacky. This is because we are accessing and manipulating the Shadow DOM in ways that it wasn't really intended. If we were to rewrite our library today, there are other ways we'd solve these issues which would respect more of the concept of the Shadow DOM and its purpose. What we do here and now may complicate a future where we have different APIs to support this vision of what support would ideally look like.
+
+
+## Backwards Compatibility Analysis
+
+This is a backwards compatible change, we should just be extending functionality, not breaking any of it.
+
+## Alternatives
+
+Unknown, haven't done research here yet.
+
+## Open Questions
+
+* How to actually define the limitations of our support? See Introduction, it's missing a final sentence with this information.
+
+## Help Needed
+
+The biggest help we can receive is tests, either in the form of unit tests or in the form of examples of real life applications/setups that we can turn into unit tests. The more tests we have, the less likely we will break anything moving forward after the initial effort is complete.
```

Comment:
We want to avoid maintaining a fork so we will try to do our best here. Right now we use patches which are a nightmare during upgrades.

Reply:
I'm not sure what this has to do with the "help needed" section? Are you just saying it will be hard to create tests and examples?

Reply:
Sorry for not being clear. I meant we will like to help here as much as possible so that we do not have to maintain a fork. At the end of the day, we want shadow DOM support to land up here.